### PR TITLE
Move from `pkg_resources` to `importlib.resources`

### LINKED ISSUE #12508: Move from `pkg_resources` to `importlib.resources`
The documentation for [`pkg_resources`](https://setuptools.pypa.io/en/latest/pkg_resources.html) declares the package to be deprecated:

> Use of pkg_resources is discouraged in favor of [importlib.resources](https://docs.python.org/3/library/importlib.html#module-importlib.resources), [importlib.metadata](https://docs.python.org/3/library/importlib.metadata.html), and their backports ([importlib_resources](https://pypi.org/project/importlib_resources), [importlib_metadata](https://pypi.org/project/importlib_metadata)). Please consider using those libraries instead of pkg_resources.

It would be nice (though by no means critically important) to follow this advice. The standalone backport module`importlib_resources` provides a [migration guide](https://importlib-resources.readthedocs.io/en/latest/migration.html).

Grepping the source tree for `pkg_resources`, we currently use:
- pkg_resources.resource_filename
- pkg_resources.resource_stream
- pkg_resources.parse_version

The first two are covered by the migration guide. The third can be replaced by `packaging.version.Version`. We already depend on packaging >= 16.1, and AFAICS from [its changelog](https://packaging.pypa.io/en/latest/changelog.html) the Version type and its functionality we use have been present since `packaging`'s initial release.

Once this and #12478 is done, we can remove the development dependency on `types-setuptools`.

Task list:

- [ ] Migrate uses of resource_filename to importlib.resources
- [ ] Migrate uses of resource_stream to importlib.resources
- [ ] Migrate uses of parse_version to packaging.version.Version
- [ ] Remove dev dependency on setuptools and update poetry lockfile (`poetry lock --no-update`)
- [ ] Run linter script
- [ ] Run tests (in CI for a PR)

### PR DESCRIPTION
**Pull request by vheuken**
Resolves #12508 

The issue also states that `setuptools` needs to be removed as a dev dependency. However, in `tox.ini`, there's comments explaining that it is required by another library. Is this comment outdated? 

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by DMRobertson** [comment]
Many thanks for taking this on @vheuken!

Some of the tests are failing on older versions of Python. This is because the porting instructions assume you're using a new version of `importlib.resources` which has the `files()` function. Unfortunately that was only added in Python 3.9. Sorry I hadn't anticipated this!

Fixing this requires a little more work. I think the easiest thing to do would be to try and use the functions in older versions of `importlib.resources` instead of `files()`. See [here](https://docs.python.org/3/library/importlib.html?highlight=importlib%20resources#importlib.resources.open_binary), and look for the ones that _don't_ say "added in version 3.9".

Does that make sense? Do you think you'd be able to try that? Feel free to fire off any questions, either way.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
(Oops, I should have written [that comment](https://github.com/matrix-org/synapse/pull/12542#issuecomment-1108896144) as a review, to take it out of Synapse's review queue.)

File: synapse/config/oembed.py [hunk 1]
```diff
@@ -58,7 +58,9 @@ def _parse_and_validate_providers(
         # Whether to use the packaged providers.json file.
         if not oembed_config.get("disable_default_providers") or False:
             providers = json.load(
-                pkg_resources.resource_stream("synapse", "res/providers.json")
+                importlib.resources.files("synapse")
+                .joinpath("res/providers.json")
+                .open("rb")
             )
```
Comment (clokep) [inline synapse/config/oembed.py:64]: I think we probably want to use a context manager here to ensure the file gets closed, something like:

```python
providers_path = importlib.resources.files(...).joinpath(...)
with open(providers_path, "rb") as f:
    providers = json.load(f)
```

**Comment by vheuken** [comment]
@DMRobertson Thanks for that. I opted to use `importlib.resources.path`, which fixed the integration tests failing, but now it looks like it fails with versions of Python older than 3.10. I'll try to get 3.9 on my local machine and test against both versions before pushing again.

**Comment by DMRobertson** [comment]
We discussed this in #synapse-dev and came to the conclusion that the change described in #12508 didn't seem to be working well in practice.

- Synapse assumes that its resources are available in a directory on the filesystem.
- importlib.resources emphasises that packages' resources may not be available as files on disk. E.g. they might be compressed in a zip, or fetched on-demand over the network somehow.
- For that reason it's awkward to get our hands on the path to the templates directory in the way that Synapse currently does. It felt like we were fighting `importlib.resources`!
- `importlib.resources.files` seemed to work okay, but isn't present in older pythons. We'd need to use the backport `importlib_resources`---but that doesn't seem to be worth the effort given the friction.

Given all that, I think it's best to close this PR and the original issue too.

@vheuken thanks again for your efforts here---you've uncovered something that we hadn't grasped at first!

**Pull request closed without merging**